bank/api/middleware.go

```diff
@@ -0,0 +1,167 @@
+package api
+
+import (
+	"encoding/json"
+	"github.com/gorilla/context"
+	"github.com/iliyanmotovski/bankv2/bank/domain"
+	"log"
+	"net/http"
+	"time"
+)
+
+func CheckValidSessionOnLoginAndLogoutMiddleware(sessStore domain.SessionStore) func(http.Handler) http.Handler {
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var u domain.User
+			const loginUrl = "/users/login"
+			const logoutUrl = "/users/logout"
+
+			json.NewDecoder(r.Body).Decode(&u)
+			cookie, err := r.Cookie("session")
+			if err != nil {
+				cookie = &http.Cookie{Value: "rfBd56ti2SMtY"}
+			}
+
+			session, ok := sessStore.FindSessionAvailableAt(cookie.Value, time.Now().Local())
+			if ok {
+				switch r.URL.String() {
+				case loginUrl:
```

Comment:
I'm not sure that you need these url checks.

```golang
CookieBasedSecurity(sessionStore domain.SessionStore) 
```

Keep in mind that you can attach middleware to specific endpoint, so you just don't have to attach this middleware for login and logout handlers.